### PR TITLE
added logic to instant shutdown strategy with withdraw money to vault

### PR DESCRIPTION
**Pull request by wer32dr**
- added logic to instant shutdown strategy with withdraw money to vault without waiting of the next harvest cycle
- fixed issue with potential underlying protocol cheating

### CODE REVIEW

File: packages/contracts/src/strategies/BaseStrategy.sol [hunk 1]
```diff
@@ -335,6 +338,18 @@ abstract contract BaseStrategy is
 
     /// @dev Fallback function that is called when the health check fails.
     function healthCheckFailedFallback() internal virtual override{
-        shutdown();
+        if (!paused()) {
```
Comment (LeoVS09) [inline packages/contracts/src/strategies/BaseStrategy.sol:341]: Better add there comment
```suggestion
        //Prevent cycle of failed health checks on shutdown 
        if (!paused()) {
```

File: packages/contracts/src/strategies/BaseStrategy.sol [hunk 2]
```diff
@@ -335,6 +338,18 @@ abstract contract BaseStrategy is
...
+            shutdown();
+            _work();
+        }
+    }
+
+    /// @notice Returns free balance of the strategy.
+    function _underlyingBalance()
```
Comment (LeoVS09) [inline packages/contracts/src/strategies/BaseStrategy.sol:348]: Let's use `_freeAssets()` simular to what we have for the Vault
https://github.com/eonian-core/farm/blob/staging/packages/contracts/src/Vault.sol#L261-L263

File: packages/contracts/test/unit/ApeLendingStrategy.t.sol [hunk 3]
```diff
@@ -800,4 +831,39 @@ contract ApeLendingStrategyTest is TestWithERC1820Registry {
...
+
+        // check if shutdown triggered
+        vm.expectEmit(true, true, true, true);
+        strategy.emitHealthCheckTriggered(2);
+
+        cToken.setLoss(loss);
+        strategy.callWork();
+
+        // check if not lost money returned back to vault after shutdown the strategy
+        assertEq(vault.freeAssets(), amount - loss);
```
Comment (LeoVS09) [inline packages/contracts/test/unit/ApeLendingStrategy.t.sol:867]: Also need check that strategy paused and nothing left on strategy balance

File: packages/contracts/src/strategies/BaseStrategy.sol [hunk 4]
```diff
@@ -266,11 +267,13 @@ abstract contract BaseStrategy is
...
-            loss = outstandingDebt - amountFreed;
-        } else if (amountFreed > outstandingDebt) {
-            profit = amountFreed - outstandingDebt;
+        _liquidateAllPositions();
+        uint256 freeBalance = _underlyingBalance();
+
+        if (freeBalance < outstandingDebt) {
+            loss = outstandingDebt - freeBalance;
+        } else if (freeBalance > outstandingDebt) {
+            profit = freeBalance - outstandingDebt;
```
Comment (LeoVS09) [inline packages/contracts/src/strategies/BaseStrategy.sol:276]: Please add test for this logic change

File: packages/contracts/src/strategies/BaseStrategy.sol (hunk 2 above)
Comment (LeoVS09) [reply packages/contracts/src/strategies/BaseStrategy.sol:348]: And also need simple test for this method

File: packages/contracts/src/strategies/CTokenBaseStrategy.sol [hunk 5]
```diff
@@ -151,9 +152,14 @@ abstract contract CTokenBaseStrategy is ICInterestRate, BaseStrategy {
...
-        emit WithdrawnFromProtocol(amountToRedeem, sharesBefore, deposits, sharesAfter, underlyingAfter);
 
-        return amountToRedeem;
+        // underlying protocol can return less than calculated above, so this check is required to avoid manipulations
+        uint256 balanceAfter = _underlyingBalance();
+        uint256 redeemedAmount = balanceAfter - balanceBefore;
+
+        emit WithdrawnFromProtocol(amountToRedeem, sharesBefore, deposits, sharesAfter, underlyingAfter, redeemedAmount);
+
+        return redeemedAmount;
```
Comment (LeoVS09) [inline packages/contracts/src/strategies/CTokenBaseStrategy.sol:162]: Please add test for this logic change

**Review by LeoVS09 — COMMENTED** [review]
Great work, only few tests missing

File: packages/contracts/src/strategies/BaseStrategy.sol [hunk 6]
```diff
@@ -168,8 +171,6 @@ abstract contract BaseStrategy is
             totalDebt,
             gasCost
         );
-
-        emit Harvested(profit, loss, debtPayment, outstandingDebt);
```
Comment (solovev) [inline packages/contracts/src/strategies/BaseStrategy.sol:172]: Why was this moved? "Harvested" event won't be emitted if the healthcheck failed and transaction reverted

File: packages/contracts/src/strategies/CTokenBaseStrategy.sol (hunk 5 above)
Comment (solovev) [reply packages/contracts/src/strategies/CTokenBaseStrategy.sol:162]: Do we really need this change?
`balanceAfter - balanceBefore` is not a redeemed amount in fact. What if there was some amount of assets on the contrac already? Or, someone send some to the contract address, so then it will produce wrong value. 

Redeemed amount is the actual amount of assets that was withdrawed from the underlying contract (ApeSwap), i.e. `redeemedAmount` = `amountToRedeem` as it was before (otherwise transaction will be reverted)

Comment (LeoVS09) [reply packages/contracts/src/strategies/CTokenBaseStrategy.sol:162]: This is correct logic, just what we requested to redeem not equal what was redeemed at the end. We need not trust third party contracts for this and check by self

File: packages/contracts/src/strategies/BaseStrategy.sol (hunk 1 above)
Comment (wer32dr) [reply packages/contracts/src/strategies/BaseStrategy.sol:342]: added

File: packages/contracts/test/unit/ApeLendingStrategy.t.sol (hunk 3 above)
Comment (wer32dr) [reply packages/contracts/test/unit/ApeLendingStrategy.t.sol:883]: added

File: packages/contracts/src/strategies/BaseStrategy.sol (hunk 4 above)
Comment (wer32dr) [reply packages/contracts/src/strategies/BaseStrategy.sol]: I'm out of ideas on how to test it properly in the current logic. However this part of code has already covered with other tests including proper branching according to coverage report

File: packages/contracts/src/strategies/CTokenBaseStrategy.sol (hunk 5 above)
Comment (wer32dr) [reply packages/contracts/src/strategies/CTokenBaseStrategy.sol]: added additional fraud test

File: packages/contracts/src/strategies/BaseStrategy.sol (hunk 6 above)
Comment (wer32dr) [reply packages/contracts/src/strategies/BaseStrategy.sol:172]: The reason for this change is recursive logic, in the case of not changing the place of emit logic this recursive logic will emit the event from the shutting down strategy first and only after it emits the event from the very initial withdrawal with the loss which is not correct order and it is better to emit it only once by the end of this method.

I thought about passing additional parameters or using extra variables (extra cost) but all those tricks look bad. Do you have an idea how to solve this issue effectively?

Comment (solovev) [reply packages/contracts/src/strategies/BaseStrategy.sol:172]: nah, maybe it's ok, if healthcheck failed, that means no haverst was performed, so let's leave it as is

**Review by solovev — APPROVED** [review]

File: packages/contracts/src/strategies/BaseStrategy.sol (hunk 4 above)
Comment (LeoVS09) [reply packages/contracts/src/strategies/BaseStrategy.sol]: Okay, let's keep existing tests
